### PR TITLE
Reorg logical flow in train

### PR DESCRIPTION
**Pull request by chengfx**
I have reorged the logical flow in train.py in order to better readability, scalability and robustness. It is the first work to add encoding cache mechanism. I have tested the regression & classification & Chinese tasks locally and still need more tests from folks 😄 because there are lots of changes in this PR.
current logical flow is 
	
1. Init
    1. Conf 
    2. Problem
    3. other
        1. finetune
2. Cache verification (Use cache == True)
    1. check
			1. Cache conf
			2. Problem
			3. Embedding
			4. Encoding
    2. load
3. Data preprocessing (Use cache == False or cache verification == false)
		1. build dictionary
		2. encoding
4. Environment Preparing
    1. Cache Save (Use cache == True)
			1. create dir
			2. conf
			3. problem
			4. embedding
			5. encoding
   2. Backup
5. Train phase
    1. init
			1. model
			2. loss
			3. optimizer
    2. train
    3. test

### CODE REVIEW

**Comment by ljshou** [comment]
牛🐮辛苦

File: problem.py [hunk 1]
```diff
@@ -50,9 +48,19 @@ def __init__(self, problem_type, input_types, answer_column_name=None, lowercase
             same_length:
             with_bos_eos: whether to add bos and eos when encoding
         """
-        self.lowercase = lowercase
 
-        self.input_dicts = dict()
+        # init
```
Comment (woailaosang) [inline problem.py]: readability is not good, I think. Strongly suggest enumerate every task.

Comment (chengfx) [reply problem.py]: Hi, @woailaosang  Readability is just special for train logical flow here. I made less works on other modules. Yeah, enumerate every task is a good idea . But I think we need to optimize the code to reduce the repeated codes and logic. Otherwise you have to modify every place if you want to make some changes

File: .gitignore [hunk 2]
```diff
@@ -2,5 +2,7 @@
 *~
 *.pyc
 *.cache*
+*.vs*
```
Comment (woailaosang) [inline .gitignore]: what's the directory of '.vs'?

Comment (chengfx) [reply .gitignore]: it's just the configuration of vs code, not related to this project 😄

File: train.py [hunk 3]
```diff
@@ -21,199 +21,204 @@
...
+    if problem.output_dict:
+        logging.debug("Problem target cell dict: %s" % (problem.output_dict.cell_id_map))
+    
+    # train phase
+    ## init 
+    ### model
+    vocab_info, initialize = None, False
+    if not conf.pretrained_model_path:
+        vocab_info, initialize = get_vocab_info(problem, emb_matrix), True
+    print(initialize)
```
Comment (woailaosang) [inline train.py]: This 'print' line needs to be deleted, I think.

File: train.py [hunk 4]
```diff
@@ -21,199 +21,204 @@
...
-                    logging.info('Old cache is deleted')
-
-            if not os.path.exists(conf.cache_dir):
-                os.makedirs(conf.cache_dir)
-                shutil.copy(params.conf_path, os.path.join(conf.cache_dir, 'conf_cache.json'))
-
-        # first time training, load problem from cache, and then backup the cache to model_save_dir/.necessary_cache/
-        if conf.use_cache and os.path.isfile(conf.problem_path):
+    def load(self, conf, problem, emb_matrix):
+        # load dictionary when (not finetune) and (cache invalid)
```
Comment (woailaosang) [inline train.py]: load dictionary when (not finetune) and (cache **valid**)

Comment (chengfx) [reply train.py]: thanks, done

**Comment by ljshou** [comment]
there are conflict on these files: Conflicting files
problem.py 
train.py

**Comment by chengfx** [comment]
> there are conflict on these files: Conflicting files
> problem.py
> train.py

done
